open-apparel-registry/open-apparel-registry
# Scheduled check_api_limits invocation with Step Functions

## Overview

The following changes were made to support the scheduled invocation of the `check_api_limits` management command:

- Create supporting IAM resources (Step Functions service role w/ policy and CloudWatch Events service role).
- Create a Step Functions state machine that will run the App CLI ECS task.
- Catch and surface failures to Rollbar via a Lambda function.
- Create a CloudWatch Events Rule that triggers the state machine with a configurable schedule.

Resolves #1165 

## Testing Instructions

- Select **Step Functions** → **State machines** → **stateMachineStagingAppCLI** .
- See that the CloudWatch Events Rule is invoking the task every hour:
   ![image](https://user-images.githubusercontent.com/1774125/103237922-280dca80-4917-11eb-9b83-a9ea158117a7.png)
- Select **Start Execution**.
  - **Input**: 
  ```json
  {
    "commands": [
      "check_api_limits_asdfghjkl"
    ]
  }
  ```
- Wait for a Rollbar message to appear in Slack.
   ![image](https://user-images.githubusercontent.com/1774125/103237593-0a8c3100-4916-11eb-8705-7b281a81d83a.png)

In order to view specific log output, you'll need to find the task in the ECS console or find the CloudWatch log stream. Although, if `check_api_limits` is successful, sometimes [there isn't any](https://github.com/open-apparel-registry/open-apparel-registry/pull/1153#issue-511568953) log output.

## Checklist
